ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -81,74 +82,116 @@ tests =
+    -- Generate PublicRootPeers
+    --
+    publicRootRelays <- vectorOf relaysNumber arbitrary
+
+    let publicRootPeers = tagRelays publicRootRelays
+
+        publicRootDomains = [ domain
+                            | RelayAccessDomain domain _ <- publicRootPeers ]
+
+    publicRootDomainIPs <- blocks ipsPerDomain
+            -- Modules under test do not differ by IP version so we only
             -- generate IPv4 addresses.
-            <$> vectorOf (ipsPerDomain * length domains)
+            <$> vectorOf (ipsPerDomain * length publicRootDomains)
                          (IPv4 . toIPv4w <$> arbitrary)
 
-    let dnsMap = Map.fromList $ zip domains ips
+    let prpDNSMap = Map.fromList $ zip publicRootDomains publicRootDomainIPs
 
     return (MockRoots {
-      mockLocalRootPeers = localRootPeers,
-      mockDNSMap         = dnsMap
+      mockLocalRootPeers        = localRootPeers,
+      mockLocalRootPeersDNSMap  = lrpDNSMap,
+      mockPublicRootPeers       = publicRootPeers,
+      mockPublicRootPeersDNSMap = prpDNSMap
     })
   where
+    tagRelays relays =
+      zipWith
+        (\tag rel
+          -> case rel of
+               RelayAccessDomain domain port
+                 -> RelayAccessDomain (domain <> (pack . show) tag) port
+               x -> x
+        )
+        [(0 :: Int), 1 .. ]
+        relays
+
     blocks _ [] = []
     blocks s l  = take s l : blocks s (drop s l)
 
 instance Arbitrary MockRoots where
     arbitrary = genMockRoots
-    shrink roots@MockRoots { mockLocalRootPeers, mockDNSMap } =
-      [ roots { mockLocalRootPeers = peers}
-      | peers <- shrinkList (const []) mockLocalRootPeers
-      ]
-      ++
-      [ roots { mockDNSMap = Map.fromList dnsMap }
-      | dnsMap <- shrinkList (const []) (Map.assocs mockDNSMap)
+    shrink roots@MockRoots { mockLocalRootPeers
+                           , mockLocalRootPeersDNSMap
+                           , mockPublicRootPeers
+                           , mockPublicRootPeersDNSMap
+                           } =
+      [ roots { mockLocalRootPeers        = lrp
+              , mockLocalRootPeersDNSMap  = lrpDNSMap
+              }
+      | lrp <- shrinkList (const []) mockLocalRootPeers,
+        let lrpDomains =
+              Set.fromList [ domain
+                           | RelayAccessDomain domain _
+                              <- concatMap (Map.keys . snd) lrp ]
+            lrpDNSMap  = Map.restrictKeys mockLocalRootPeersDNSMap
+                                          lrpDomains
+      ] ++
+      [ roots { mockPublicRootPeers       = prp
+              , mockPublicRootPeersDNSMap = prpDNSMap
+              }
+      | prp <- shrinkList (const []) mockPublicRootPeers,
+        let prpDomains = Set.fromList [ domain
+                                      | RelayAccessDomain domain _ <- prp ]
+            prpDNSMap  = Map.restrictKeys mockPublicRootPeersDNSMap
+                                          prpDomains
```

Comment:
This looks good, could you say what was the problem?

Reply:
The problem in t his case is that the DNSMap needs to have a subset of the domains. In the previous version of the code the DNSMap would shrink independently of the domains and make the shrinking loop indefinitely because the tests would always fail